sigmastate/src/test/scala/sigmastate/UncheckedTreeSpecification.scala

```diff
@@ -0,0 +1,104 @@
+    def cOrUncheckedNode(challenge: Challenge) = COrUncheckedNode(challenge, List.empty[UncheckedSigmaTree])
+    def cThresholdUncheckedNode(challenge: Challenge) = CThresholdUncheckedNode(challenge, List.empty[UncheckedSigmaTree], 0, None)
+
+    val emptyChallenge = Challenge @@ Array.emptyByteArray
+    val size3Challenge = Challenge @@ ErgoAlgos.decodeUnsafe("010203")
+    val size4Challenge = Challenge @@ ErgoAlgos.decodeUnsafe("01020304")
+    val bigChallenge = Challenge @@ ErgoAlgos.decodeUnsafe("c6429b70f4926a3ba1454f1aec116075f9e9fbe8a8f72114")
+
+    val smallSizeExamples =
+      Table(
+        "uncheckedTree",
+        uncheckedSchnorr(emptyChallenge),
+        uncheckedSchnorr(size3Challenge),
+        uncheckedDiffieHellmanTuple(emptyChallenge),
+        uncheckedDiffieHellmanTuple(size3Challenge),
+        cAndUncheckedNode(emptyChallenge),
+        cAndUncheckedNode(size3Challenge),
+        cOrUncheckedNode(emptyChallenge),
+        cOrUncheckedNode(size3Challenge),
+        cThresholdUncheckedNode(emptyChallenge),
+        cThresholdUncheckedNode(size3Challenge)
+      )
+
+    forAll(smallSizeExamples) { uncheckedTree =>
+      uncheckedTree.challengeOptimizedHash shouldBe Arrays.hashCode(uncheckedTree.challenge)
+    }
+
+    val bigSizeExamples = Table(
+      "uncheckedTree",
+      uncheckedSchnorr(size4Challenge),
+      uncheckedSchnorr(bigChallenge),
+      uncheckedDiffieHellmanTuple(size4Challenge),
+      uncheckedDiffieHellmanTuple(bigChallenge),
+      cAndUncheckedNode(size4Challenge),
+      cAndUncheckedNode(bigChallenge),
+      cOrUncheckedNode(size4Challenge),
+      cOrUncheckedNode(bigChallenge),
+      cThresholdUncheckedNode(size4Challenge),
+      cThresholdUncheckedNode(bigChallenge)
+    )
+
+    forAll(bigSizeExamples) { uncheckedTree =>
+      uncheckedTree.challengeOptimizedHash shouldBe Ints.fromByteArray(uncheckedTree.challenge)
+    }
+
+    val equalHashCodes = Table(
+      "uncheckedTree1",
+      uncheckedSchnorr(emptyChallenge),
+      uncheckedSchnorr(size3Challenge),
+      uncheckedSchnorr(size4Challenge),
+      uncheckedSchnorr(bigChallenge),
+      uncheckedDiffieHellmanTuple(emptyChallenge),
+      uncheckedDiffieHellmanTuple(size3Challenge),
+      uncheckedDiffieHellmanTuple(size4Challenge),
+      uncheckedDiffieHellmanTuple(bigChallenge),
+      cAndUncheckedNode(emptyChallenge),
+      cAndUncheckedNode(size3Challenge),
+      cAndUncheckedNode(size4Challenge),
+      cAndUncheckedNode(bigChallenge),
+      cOrUncheckedNode(emptyChallenge),
+      cOrUncheckedNode(size3Challenge),
+      cOrUncheckedNode(size4Challenge),
+      cOrUncheckedNode(bigChallenge),
+      cThresholdUncheckedNode(emptyChallenge),
+      cThresholdUncheckedNode(size3Challenge),
+      cThresholdUncheckedNode(size4Challenge),
+      cThresholdUncheckedNode(bigChallenge)
+    )
+
+    forAll(equalHashCodes) { uncheckedTree =>
+      uncheckedTree.hashCode shouldBe uncheckedTree.hashCode  
```

Comment:
what are you testing here?

Reply:
Basically determinism on hashCode function. Changed in newer version.